File: Adapter/Adapter.cpp
#include<iostream>
using namespace std;

/*
* How many characters in Adapter pattern?
  1.Adapter, designed to matches Adaptee implements with the Target-interface
  2.Target, raise a request
  3.Adaptee, contains implements which does not fit with Target-interface
* Class-Adapter and Object-Adapter are showed in this .cpp file.
*/

class Adaptee {
public:
	void SpecificImplement (){
		cout << "It's an implement of Adaptee class, cannot be called directly by Target" << endl;
	}
};
class Target {
public:
	virtual void Implement() {
		cout << "Target Implement needed." << endl;
	}
};

//Class Adapter
class ClassAdapter : public Target, private Adaptee{
public:
	void Implement() {
		SpecificImplement();
		cout << "Adapter : Now it matches me! ___by ClassAdapter" << endl;
	}
};

//Object Adapter
class ObjectAdapter : public Target {
public:
	ObjectAdapter(Adaptee* adaptee) {
		_adaptee = adaptee;
	}
	void Implement() {
		_adaptee->SpecificImplement();
		cout << "Adapter : Now it matches me! ___by ObjectAdapter" << endl;
	}
private:
	Adaptee* _adaptee;
};

int main() {
	Target* tClass = new ClassAdapter();
	tClass -> Implement();

	cout << "**************************" << endl;

	Target* tObj = new ObjectAdapter(new Adaptee());
	tObj -> Implement();
}

File: Adapter/Adapter.cs
using System;
using System.Dynamic;

/*
* How many characters in Adapter pattern?
  1.Adapter, designed to matches Adaptee implements with the Target-interface
  2.Target, raise a request
  3.Adaptee, contains implements which does not fit with Target-interface
* Class-Adapter and Object-Adapter are showed in this .cpp file.
*/

class Adaptee
{
    public void SpecificImplement()
    {
        Console.WriteLine("It's an implement of Adaptee class, cannot be called directly by Target");
    }
};
interface ITarget
{
    void Implement();
};

//Class Adapter
class ClassAdapter : Adaptee, ITarget
{
    public void Implement()
    {
        SpecificImplement();
        Console.WriteLine("Adapter : Now it matches me! ___by ClassAdapter");
    }
};

//Object Adapter
class ObjectAdapter : ITarget
{
    public ObjectAdapter(Adaptee adaptee)
    {
        _adaptee = adaptee;
    }
    public void Implement()
    {
        _adaptee.SpecificImplement();
        Console.WriteLine("Adapter : Now it matches me! ___by ObjectAdapter");
    }
    private Adaptee _adaptee;
};

class Client
{
    static void Main()
    {
        ITarget tClass = new ClassAdapter();
        tClass.Implement();

        Console.WriteLine("**************************");

        ITarget tObj = new ObjectAdapter(new Adaptee());
        tObj.Implement();
    }
}